### DISCUSSION TITLE
New Plugin Trouble

### DISCUSSION THREAD

**Discussion by Eslifkin23** [jspsych/jsPsych#982] [Q&A] [answered]
Hi,

I have an experiment that runs 3 plugins and then finishes.  The problem I'm having is with the 3rd (and last) plugin.  Basically I just need this plugin to save data created by a function without actually showing anything on the screen.  Currently when I run the experiment, it runs the first 2 plugins no problem, then shows a blank screen and doesn't proceed.  I assume this blank screen is the 3rd plugin running and not stopping.  I want this plugin to timeout after a very short time and then the experiment should continue (it should end and show all the data).  Here is the code for my 3rd plugin:

```
jsPsych.plugins['vcr-deg2px'] = (function () {

    var plugin = {};

    plugin.info = {
        name: 'vcr-deg2px',
        parameters: {
        }
    }

    plugin.trial = function (display_element, trial) {

        function deg2px(deg) {
            var lpdValues = jsPsych.data.get().select('lpd').values
            var lpd = lpdValues[lpdValues.length - 1] 
            var distanceValues = jsPsych.data.get().select('viewDistance').values
            var distance = distanceValues[distanceValues.length - 1]
            return Math.round(Math.tan((deg/2)*(Math.PI/180))*2*distance*lpd);
        }

            jsPsych.finishTrial({
                deg2px: deg2px(deg)
                });
    }

    return plugin;

})();
```

All I need is for the deg2px function to run and then save the output of the function and have the plugin end after 200ms.  I'm not sure how to make the trial/plugin end after a set amount of time and I don't know if the data is saving since it doesn't get past the blank screen on the plugin.

Any help is appreciated!
Elisabeth

**Comment by jodeleeuw** [comment, marked as answer]
Hi Elisabeth,

In the `finishTrial` part of this plugin it looks like you are calling `deg2px(...)` with the `deg` variable as an argument but this variable is not defined anywhere. If that variable is in the trial data then you can refer to it is `trial.deg`.

To add a 200ms delay, you can use `setTimeout()`. Note that this will approximate a 200ms delay, but it may be off by one or two frames (16.67ms).

```js
setTimeout(function(){
	jsPsych.finishTrial({
		deg2px: deg2px(deg)
	});
}, 200)
```

**Reply by Eslifkin23** [reply]
From what I understand of the deg2px function (I didn't create it), I need to specify a value for the first (deg) value and it will then use that value in the second instance of (deg).  For example,
```
        function deg2px(deg=.78) {
            var lpdValues = jsPsych.data.get().select('lpd').values
            var lpd = lpdValues[lpdValues.length - 1] 
            var distanceValues = jsPsych.data.get().select('viewDistance').values
            var distance = distanceValues[distanceValues.length - 1]
            return Math.round(Math.tan((deg/2)*(Math.PI/180))*2*distance*lpd);
        }
```
I'm not positive if this is correct, but that's my understanding of it.  I don't know if this reconciles your above statement of referring to it as trial.deg

In regards to the timeout, the exact time doesn't matter really, so approximately 200ms should be fine.  I have added your above code, but it doesn't seem to be doing anything for me.

I am trying to push my updates to my online repo, but I'm having trouble connecting right now, so I've attached a zip folder of my entire experiment if that helps you at all.

Thank you!
[vcr-test-deg2px.zip](https://github.com/jspsych/jsPsych/files/5086851/vcr-test-deg2px.zip)

**Reply by jodeleeuw** [reply]
I'm talking about this part of the code:

```js
jsPsych.finishTrial({
	deg2px: deg2px(deg)
});
```

This is where `deg` is currently undefined. What's the angle that you are hoping to convert to pixels?

**Reply by Eslifkin23** [reply]
Okay, that also makes sense.  So was the original code okay where the function just had deg2px(deg)?  The angle I want to convert is .78.  Would it look like deg2px: deg2px(.78)?

**Reply by Eslifkin23** [reply]
So assuming what I said above is correct, this works!  When I change it to 
```        setTimeout(function(){
            jsPsych.finishTrial({
                deg2px: deg2px(.78)
            });
        }, 200)
```
the plugin finishes and shows me the data at the end!  If what I've done isn't what you were trying to explain, please let me know!

**Reply by jodeleeuw** [reply]
Yep, that's exactly what I meant. 

I'm imagining that you'll want to use different values of `deg` each time you run the plugin. In that case you can use:

```js
jsPsych.finishTrial({
	deg2px: deg2px(trial.deg)
});
```

and in your timeline you can set the value of `deg`

```js
var convert = {
	type: 'vcr-deg2px',
	deg: .78
}
```

**Reply by Eslifkin23** [reply]
Great, thank you!! In this case, I'm actually going to be using the same value for `deg` each time, as I want all stimuli to be sized to a specific degree of visual angle regardless of people's monitor size and resolution.
